### PR TITLE
[BEAM-13041][Playground] Prepare files with java code

### PR DESCRIPTION
**Pull request by AydarZaynutdinov**
Add Prepare step as one of the executor's steps;
Add Java preparation methods to remove "package ..." string from file;
Add Java preparation methods to change "public class" strings to "class" string into file;

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/internal/preparators/preparator.go [hunk 1]
```diff
@@ -13,15 +13,9 @@
...
 
-// NewGoExecutor creates an executor with Go specifics
-func NewGoExecutor(fs *fs_tool.LifeCycle, javaValidators *[]validatorWithArgs) *Executor {
-	return nil
-}
-
-// GetGoValidators return validators methods that needed for Go file
-func GetGoValidators() *[]validatorWithArgs {
-	return nil
+type Preparator struct {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/preparator.go:18]: Did we plan to use a common struct for validators and preparators, didn't we?

File: playground/backend/internal/preparators/java_preparators.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		log.Printf("Preparation: Error during create new temporary file, err: %s\n", err.Error())
+		return err
+	}
+	defer tmp.Close()
+
+	reg := regexp.MustCompile(pattern)
+	scanner := bufio.NewScanner(file)
+	firstLine := true
+	for scanner.Scan() {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/java_preparators.go:102]: Very complex structure, double nested if and nested for. Can we simplify that?

File: playground/backend/internal/preparators/java_preparators.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// createTempFile creates temporary file near with originalFile
+func createTempFile(originalFilePath string) (*os.File, error) {
+	// all folders which are included in filePath
+	filePathSlice := strings.Split(originalFilePath, "/")
+	fileName := filePathSlice[len(filePathSlice)-1]
+
+	// find parent folder for file
+	folderPath := "/"
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/java_preparators.go:72]: what does it mean? DONT use hardcoded paths

File: playground/backend/internal/preparators/java_preparators.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	filePathSlice = filePathSlice[:len(filePathSlice)-1]
+	for _, folder := range filePathSlice {
+		if folder == "" {
+			continue
+		}
+		folderPath = filepath.Join(folderPath, folder)
+	}
+	return os.Create(folderPath + "/tmp_" + fileName)
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/java_preparators.go:80]: If it's const - it should be moved to const

File: playground/backend/internal/preparators/java_preparators.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		log.Printf("Preparation: Error during remove public modification for class, err: %s\n", err.Error())
+		return err
+	}
+	return nil
+}
+
+// removeAdditionalPackage removes packages from java file
+func removeAdditionalPackage(args ...interface{}) error {
+	filePath := args[0].(string)
+	if err := replace(filePath, `package ([\w]+\.)+[\w]+;`, ""); err != nil {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/java_preparators.go:58]: I should go deep into the code and find method implementation to change reg exp, shouldn't I?

File: playground/backend/internal/preparators/java_preparators.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		Args:    preparatorArgs,
+	}
+	preparation := []Preparator{publicClassModification, additionalPackage}
+	return &preparation
+}
+
+// removePublicClassModifier removes public modification for class from java file
+func removePublicClassModifier(args ...interface{}) error {
+	filePath := args[0].(string)
+	if err := replace(filePath, "public class ", "class "); err != nil {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/preparators/java_preparators.go:48]: constants should be constants

File: playground/backend/internal/preparators/java_preparators.go (hunk 2 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go:77]: Changed.

File: playground/backend/internal/preparators/java_preparators.go (hunk 3 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

File: playground/backend/internal/preparators/java_preparators.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	additionalPackage := Preparator{
+		Prepare: replace,
+		Args:    []interface{}{filePath, packagePattern, emptyStringPattern},
+	}
+	preparation := []Preparator{publicClassModification, additionalPackage}
+	return &preparation
+}
+
+// replace process file by filePath and replace all patterns to newPattern
```
Comment (KhaninArtur) [inline playground/backend/internal/preparators/java_preparators.go:52]:
```suggestion
// replace processes file by filePath and replaces all patterns to newPattern
```

File: playground/backend/internal/preparators/java_preparators.go (hunk 4 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

File: playground/backend/internal/preparators/java_preparators.go (hunk 5 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

File: playground/backend/internal/preparators/java_preparators.go (hunk 6 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

File: playground/backend/internal/preparators/java_preparators.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		matches := reg.FindAllString(line, -1)
+		for _, str := range matches {
+			line = strings.ReplaceAll(line, str, newPattern)
+		}
+		if _, err = io.WriteString(tmp, line); err != nil {
+			log.Printf("Preparation: Error during write \"%s\" to tmp file, err: %s\n", line, err.Error())
+			return err
+		}
+		newLine = true
+	}
```
Comment (KhaninArtur) [inline playground/backend/internal/preparators/java_preparators.go:93]: Let's make this block a separate function – seems like we can easily extract it and simplify the `replace` function.

File: playground/backend/internal/preparators/java_preparators.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+	if err = os.Rename(tmp.Name(), filePath); err != nil {
+		log.Printf("Preparation: Error during rename temporary file, err: %s\n", err.Error())
+		return err
+	}
+	return nil
+}
+
+// createTempFile creates temporary file near with originalFile
```
Comment (KhaninArtur) [inline playground/backend/internal/preparators/java_preparators.go:105]:
```suggestion
// createTempFile creates temporary file next to originalFile
```

File: playground/backend/internal/preparators/java_preparators.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	// all folders which are included in filePath
+	filePathSlice := strings.Split(originalFilePath, pathSeparatorPattern)
+	fileName := filePathSlice[len(filePathSlice)-1]
+
+	tmpFileName := fmt.Sprintf("%s_%s", tmpFileSuffix, fileName)
+	tmpFilePath := strings.Replace(originalFilePath, fileName, tmpFileName, 1)
+	return os.Create(tmpFilePath)
+}
+
+// addNewLine add new line to tmp file
```
Comment (KhaninArtur) [inline playground/backend/internal/preparators/java_preparators.go:116]:
```suggestion
// addNewLine adds a new line to tmp file
```

Comment (KhaninArtur) [reply playground/backend/internal/preparators/java_preparators.go:116]: Seems like this function adds a new line at the end of the file – can we mention it in the docstring?

File: playground/backend/internal/preparators/java_preparators.go (hunk 9 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

File: playground/backend/internal/preparators/java_preparators.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	"bufio"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// GetJavaPreparation return preparation methods that should be applied to Java code
```
[the remark this answers is not in the document]
Comment (AydarZaynutdinov) [reply playground/backend/internal/preparators/java_preparators.go]: Changed.

**Pull request closed without merging**